Reflog support in Nessie

ajantha-bhat opened the pull request:
Reflog backend entity support to get the complete details of what happened at the backend (similar to git reflog).

Reflog support for both non-transactional db and transactional db.
- write support (protobuf serialized entries) for all operations (commit, create/delete/assign ref, merge, transplant).

Reflog entry contents can be seen here:
https://github.com/ajantha-bhat/nessie/blob/reflog/versioned/persist/serialize/src/main/proto/persist.proto#L81-L106

Added two tables for transactional store, 
One for storing the data and one for storing the reflog head.
https://github.com/ajantha-bhat/nessie/blob/reflog/versioned/persist/tx/src/main/java/org/projectnessie/versioned/persist/tx/SqlStatements.java#L128-L164

- read support with new "reflogs" REST resource and swagger API doc with 
  - pagination support
  - startHash , endHash support 
  - Authz support


phase2 TODO: (In the follow up PR)
a. python support

Please find the benchmarks attached. Run locally on my machine.
JMH version: 1.34
VM version: JDK 11.0.11, Java HotSpot(TM) 64-Bit Server VM, 11.0.11+9-LTS-194
VM invoker: /Library/Java/JavaVirtualMachines/jdk-11.0.11.jdk/Contents/Home/bin/java
VM options:
Blackhole mode: full + dont-inline hint (auto-detected, use -Djmh.blackhole.autoDetect=false to disable)
Warmup: 2 iterations, 2000 ms each
Measurement: 3 iterations, 5000 ms each
Timeout: 10 min per iteration
Threads: 4 threads, will synchronize iterations
Benchmark mode: Throughput, ops/time
Benchmark: org.projectnessie.versioned.persist.benchmarks.CommitBench.singleBranchUnsharedKeys
Parameters: (adapter = RocksDB, tablesPerCommit = 5)

<img width="1251" alt="Screenshot 2022-01-10 at 8 46 33 PM" src="https://user-images.githubusercontent.com/5889404/148790597-c031a2fd-c1de-4fec-9eba-ebd059c7ee34.png">
<img width="1251" alt="Screenshot 2022-01-10 at 8 46 43 PM" src="https://user-images.githubusercontent.com/5889404/148790613-97d65069-f3aa-45c8-9820-d8b07a53c2ab.png">

ajantha-bhat commented on versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTransactionalDatabaseAdapter.java:
```diff
@@ -676,6 +811,14 @@ protected abstract void writeGlobalCommit(
+   * consistency checks/guarantees.
+   */
+  protected abstract void writeRefLog(NonTransactionalOperationContext ctx, RefLogEntry entry)
+      throws ReferenceConflictException;
```
Thought of having `RefLogConflictException` instead of this. But then all the backend store doesn't throw exception. As reflog writing is common for all the operations, all signatures need to be changed from backend store to REST API If I add new exception.  Hence didn't modified. Any thoughts ?

snazy reviewed (changes requested):
Overall looks okay so far.

However, I don't like that database-adapter internales are exposed (list of ref-iog-id+commit-hash pairs).
The commit message should not be stored in the ref-log at all. Having the commit-hash is good enough. If the user wants information about the commit referenced by the ref-log (e.g. `CommitMeta` or `Operations`), it can be optionally retrieved and as a pro be retrieved for all non-commit operations.

CEL filtering should go out. Ref-log is "dumb" for now - no need fo add unneeded functionality for now.


> authz support

You need a new rule, no? Only special users (think: superuser, admin, gc-service-accounts) shall be able to retrieve the ref-log. It's not a "normal user operation".


> update openapi.yml

Means what?


> transactional db support (how to get head of reflog? keeping another table will create a bottleneck for concurrent write? Also at a time only one concurrent op can be success as parent id need rebase?)

It should be an `UPDATE ref_log_pointer SET ref_log_head_id = :new_head WHERE ... AND ref_log_head_id = :previous_head` - needs to be initialized with the repo.


> txn cleanup

???



> python support

Later.


> marking row to fetch intermediate rows and parallelize them from spark job ?

???

harshm-dev commented on versioned/persist/rocks/src/main/java/org/projectnessie/versioned/persist/rocks/RocksDatabaseAdapter.java:
```diff
@@ -322,4 +327,45 @@ protected int entitySize(CommitLogEntry entry) {
+  @Override
+  protected void writeRefLog(NonTransactionalOperationContext ctx, AdapterTypes.RefLogEntry entry)
+      throws ReferenceConflictException {
+    Lock lock = dbInstance.getLock().writeLock();
```
I'm a little worried that we're acquiring the lock separately for the ref log entry. If there's a commit done, the commit and reflog operations will take locks independent of each other, if commit operation passes, whereas reflog insert fails, it's gonna leave an inconsistent state.

ajantha-bhat replied:
multi table transaction is handled by CAS operation loop present in the caller NonTransactionalDatabaseAdapter#casOpLoop.  

which will takes care of reverting all the entries when any one of the commit failed.

ajantha-bhat commented on versioned/spi/src/main/java/org/projectnessie/versioned/RefLogEntry.java:
```diff
@@ -0,0 +1,55 @@
+  Hash getCommitHash();
+
+  /** Zero, one or more ancestors of this entry, nearest parent first. */
+  List<Ancestor> getAncestors();
```
finally I need list of commit hashes for gc to avoid fetching each row. But I want last reflog id also to continue calling back the api again. So, storing both parent reflog and commit hash

snazy replied:
First, it's still an implementation detail - second, you have the commit hash in RefLogEntry in the RefLogResponse, no?

ajantha-bhat commented on versioned/spi/src/main/java/org/projectnessie/versioned/RefLogNotFoundException.java:
```diff
@@ -0,0 +1,47 @@
+import javax.annotation.Nonnull;
+
+/** Exception thrown when a reflog is not present in the store. */
+public class RefLogNotFoundException extends VersionStoreException {
```
But reflog is not mapped to reference. So, added this

ajantha-bhat commented on model/src/main/java/org/projectnessie/model/RefLogResponse.java:
```diff
@@ -0,0 +1,75 @@
+
+    String getOperation();
+
+    String getDetails();
```
I decided to keep as "Commit on reference $refName with hash $comitHash" in the details and remove commit message. So, if user need more details, they can query commit log api with the information present in the details.

snazy replied:
That information is available in the respsctive fields - no need to keep another string around.

ajantha-bhat replied:
Isn't it better for UI if details has the string? instead of they forming string only for commit operation? Rest of operations also I am providing a nice string. It is duplicate storage for some operations.

snazy replied:
No, that's actually not worth it. Every UI displays/formats it differently - also l10n i18n

ajantha-bhat replied:
@snazy: I think you replied to old message. I don't have details field now.

snazy commented on servers/services/src/main/java/org/projectnessie/services/impl/RefLogApiImpl.java:
```diff
@@ -0,0 +1,160 @@
+              entry -> {
+                ImmutableRefLogResponseEntry.Builder logEntry =
+                    RefLogResponse.RefLogResponseEntry.builder();
+                if (entry.getAncestors() != null) {
```
Still, the *list* of parents/ancestors is an implementation detail that must not be exposed.

snazy commented on versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java:
```diff
@@ -1334,4 +1336,74 @@ protected void checkExpectedGlobalStates(
+    return refLogFetcher(ctx, initial, this::fetchPageFromRefLog, RefLogEntry::getAncestors);
+  }
+
+  protected <T> Spliterator<T> refLogFetcher(
```
This one should be unified with the existing method - and not copied.

ajantha-bhat replied:
yes. done.

ajantha-bhat commented on model/src/main/java/org/projectnessie/model/RefLogResponse.java:
```diff
@@ -0,0 +1,75 @@
+    /** time in microseconds. */
+    long getOperationTime();
+
+    String getOperation();
```
this is the end result. So string. 
Storing as enum to save space.

ajantha-bhat replied:
similar to how we show hashes as string.

ajantha-bhat commented on versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java:
```diff
@@ -1334,4 +1336,74 @@ protected void checkExpectedGlobalStates(
+
+  protected <T> Spliterator<T> refLogFetcher(
+      OP_CONTEXT ctx,
+      T initial,
```
reusing the logfetcher now as I changed Ancestor to just Hash.

ajantha-bhat commented on versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java:
```diff
@@ -1334,4 +1336,74 @@ protected void checkExpectedGlobalStates(
+    if (initial == null) {
+      throw RefLogNotFoundException.forRefLogId(initialHash.asString());
+    }
+    return refLogFetcher(ctx, initial, this::fetchPageFromRefLog, RefLogEntry::getAncestors);
```
reusing the logfetcher now as I changed Ancestor to just Hash.

ajantha-bhat commented on versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java:
```diff
@@ -1334,4 +1336,74 @@ protected void checkExpectedGlobalStates(
+      Function<T, List<RefLogEntry.Ancestor>> nextPage) {
+    return new AbstractSpliterator<T>(Long.MAX_VALUE, 0) {
+      private Iterator<T> currentBatch;
+      private boolean eof;
```
reusing the logfetcher now as I changed Ancestor to just Hash.

ajantha-bhat commented on versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java:
```diff
@@ -1334,4 +1336,74 @@ protected void checkExpectedGlobalStates(
+        } else if (currentBatch == null) {
+          currentBatch = Collections.singletonList(initial).iterator();
+        } else if (!currentBatch.hasNext()) {
+          if (previous == null) {
```
reusing the logfetcher now as I changed Ancestor to just Hash.

ajantha-bhat commented on versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java:
```diff
@@ -1334,4 +1336,74 @@ protected void checkExpectedGlobalStates(
+          List<RefLogEntry.Ancestor> page = nextPage.apply(previous);
+          previous = null;
+          if (!page.isEmpty()) {
+            currentBatch = fetcher.apply(ctx, page).iterator();
```
reusing the logfetcher now as I changed Ancestor to just Hash.

ajantha-bhat commented on model/src/main/java/org/projectnessie/api/params/RefLogParams.java:
```diff
@@ -0,0 +1,169 @@
+  @Pattern(regexp = Validation.HASH_REGEX, message = Validation.HASH_MESSAGE)
+  @Parameter(
+      description =
+          "Hash on the given ref to start from (in chronological sense), the 'far' end of the reflog, returned "
```
added

ajantha-bhat commented on servers/services/src/main/java/org/projectnessie/services/impl/RefLogApiImpl.java:
```diff
@@ -0,0 +1,160 @@
+          try {
+            return script.execute(Boolean.class, ImmutableMap.of(VAR_REFLOG_ENTRY, refLogEntry));
+          } catch (ScriptException e) {
+            throw new RuntimeException(e);
```
removed CEL filter feature as it was concluded that no need.

ajantha-bhat commented on servers/rest-services/src/main/java/org/projectnessie/services/rest/RestRefLogResource.java:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  private RefLogApi resource() {
+    // TODO: need RefLogApiImplWithAuthorization.
```
added RefLogApiImplWithAuthorization

ajantha-bhat commented on model/src/main/java/org/projectnessie/api/params/RefLogParams.java:
```diff
@@ -0,0 +1,169 @@
+              + "'operationTime' (long), "
+              + "'operation' (string), 'details' (string)\n\n")
+  @QueryParam("filter")
+  private String filter;
```
removed.

ajantha-bhat commented on servers/services/src/main/java/org/projectnessie/services/impl/RefLogApiImpl.java:
```diff
@@ -0,0 +1,160 @@
+
+      if (items.size() == max + 1) {
+        return ImmutableRefLogResponse.builder()
+            .addAllLogEntries(items.subList(0, max))
```
sublist creates a view from original list, doesn't allocate new list.
And the allLogEntries is called only when the limit is reached. It is not called else.

ajantha-bhat commented on servers/services/src/main/java/org/projectnessie/services/impl/RefLogApiImpl.java:
```diff
@@ -0,0 +1,160 @@
+            .build();
+      }
+      return ImmutableRefLogResponse.builder().addAllLogEntries(items).build();
+    } catch (RefLogNotFoundException e) {
```
good point. need to discuss separately. I think reflog cannot support old store. So, I will create an issue to track it.

ajantha-bhat commented on versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTransactionalDatabaseAdapter.java:
```diff
@@ -402,9 +512,23 @@ public void initializeRepo(String defaultBranchName) {
+                commitTimeInMicros(),
+                ByteString.copyFromUtf8(
+                    String.format(
+                        "Create Default branch %s with %s", defaultBranchName, NO_ANCESTOR)));
```
I want to make it related to CREATE_REFERENCE, hence this description.

ajantha-bhat commented:
@snazy , @harshm-dev : I have handled the comments and also finished TODO which was applicable for non-txn db.
please have a look at the PR again. Thanks.

ajantha-bhat commented on model/src/main/java/org/projectnessie/model/RefLogResponse.java:
```diff
@@ -0,0 +1,75 @@
+@Schema(type = SchemaType.OBJECT, title = "RefLogResponse")
+@JsonSerialize(as = ImmutableRefLogResponse.class)
+@JsonDeserialize(as = ImmutableRefLogResponse.class)
+public interface RefLogResponse extends PaginatedResponse {
```
added.

snazy commented on servers/services/src/main/java/org/projectnessie/services/impl/RefLogApiImpl.java:
```diff
@@ -0,0 +1,106 @@
+
+  @Override
+  public RefLogResponse getRefLog(RefLogParams params) throws NessieNotFoundException {
+    int max =
```
The code here looks very much like the code to retrieve the commit log.
Can it be unified and the common parts implemented in BaseApiImpl?

ajantha-bhat replied:
The building different objects here also we don't need filter and get commit log needs reference. hence cannot combine. Not much common code left. They look similar but for different purpose.

snazy commented on versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTransactionalDatabaseAdapter.java:
```diff
@@ -853,4 +997,72 @@ protected abstract GlobalStateLogEntry fetchFromGlobalLog(
+    return currentRefLogId;
+  }
+
+  private RefLogEntry buildRefLogEntry(
```
This should live in `ABstractDatabaseAdapter` like `buildIndividualCommit` does

ajantha-bhat replied:
I intentionally didn't added there to avoid having `ReflogEntry` to `adaptertypes.ReflogEntry` conversion. If I keep it inside like `writeGlobalCommit`, I can avoid that extra conversion as I can build `adaptertypes.ReflogEntry `directly.

So, I am basically following like `writeGlobalCommit`

snazy commented on model/src/main/java/org/projectnessie/model/RefLogResponse.java:
```diff
@@ -0,0 +1,76 @@
+
+    /** Detailed message to explain the operation. */
+    @NotNull
+    String getDetails();
```
None of the implementations add any value to this property - i.e. all information in "details"  is available via the other fiels. Really no need to keep this property here or even persisted. It would be okay to add the merged/transplanted hashes as a separate, optional attribute. But I'm -1 on returning or persisting a human readable message that doesn't really add value.

snazy commented on versioned/persist/serialize/src/main/proto/persist.proto:
```diff
@@ -77,10 +77,34 @@ message GlobalStateLogEntry {
+    TRANSPLANT = 5;
+  }
+  Operation operation = 7;
+  bytes details = 8;
```
None of the implementations add any value to this property - i.e. all information in "details"  is available via the other fiels. Really no need to keep this property. It would be okay to add the merged/transplanted hashes as a separate, optional attribute. But I'm -1 on persisting a human readable message that doesn't really add value.

snazy commented on versioned/spi/src/main/java/org/projectnessie/versioned/RefLogNotFoundException.java:
```diff
@@ -0,0 +1,47 @@
+import javax.annotation.Nonnull;
+
+/** Exception thrown when a reflog is not present in the store. */
+public class RefLogNotFoundException extends VersionStoreException {
```
Any test case that verifies that this excepiton is properly thrown?

snazy commented on versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTransactionalDatabaseAdapter.java:
```diff
@@ -853,4 +997,72 @@ protected abstract GlobalStateLogEntry fetchFromGlobalLog(
   protected abstract List<GlobalStateLogEntry> fetchPageFromGlobalLog(
       NonTransactionalOperationContext ctx, List<Hash> hashes);
+
+  protected Hash writeRefLogEntry(
```
This should live in `ABstractDatabaseAdapter`

ajantha-bhat replied:
As mentioned above,
I intentionally didn't added there to avoid having ReflogEntry to adaptertypes.ReflogEntry conversion. If I keep it inside like writeGlobalCommit, I can avoid that extra conversion as I can build adaptertypes.ReflogEntry directly.

So, I am basically following like writeGlobalCommit

snazy commented on versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTransactionalDatabaseAdapter.java:
```diff
@@ -208,8 +210,21 @@ public Hash merge(
                     ctx, timeInMicros, Hash.of(pointer.getGlobalId()), Collections.emptyList());
 
+            Hash newRefLogId =
+                writeRefLogEntry(
```
This and all other call sites of `writeRefLogEntry` look tike they will be duplicated in `TxDatabaseAdapter` - better to not have duplicated code.

ajantha-bhat replied:
yeah. If I move this to abstract class, then I need to use `RefLog` instead of proto `ReflogEntry`.
So, it brings an extra conversion for each operation. 
Considering code duplication vs perf. I chose performance. 

Actually I tired adding proto dependency to abstract class. but its a circular dependency. I think some part is not designed common for txn and non-txn.

snazy commented on versioned/persist/dynamodb/src/main/java/org/projectnessie/versioned/persist/dynamodb/DynamoDatabaseAdapter.java:
```diff
@@ -117,7 +120,7 @@ public void eraseRepo() {
         b -> b.tableName(TABLE_GLOBAL_POINTER).key(globalPointerKeyMap).build());
 
-    Stream.of(TABLE_GLOBAL_LOG, TABLE_GLOBAL_LOG, TABLE_KEY_LISTS)
+    Stream.of(TABLE_GLOBAL_LOG, TABLE_COMMIT_LOG, TABLE_KEY_LISTS, TABLE_REF_LOG)
```
Nice catch!

snazy commented on versioned/persist/serialize/src/main/proto/persist.proto:
```diff
@@ -77,10 +77,34 @@ message GlobalStateLogEntry {
   repeated ContentIdWithBytes puts = 4;
 }
 
+message RefLogEntry {
```
Rename this one to avoid fully-qualified class names - clashes with `org.projectnessie.versioned.RefLogEntry` and  `org.projectnessie.model.RefLogResponse.RefLogEntry`

ajantha-bhat replied:
changed now.
proto is having RefLogEntry
RefLogResponse is having RefLogResponseEntry
spi is having RefLog

snazy commented on versioned/persist/store/src/main/java/org/projectnessie/versioned/persist/store/PersistVersionStore.java:
```diff
@@ -152,7 +154,6 @@ public Hash commit(
     }
 
     commitAttempt.commitMetaSerialized(serializeMetadata(metadata));
-
```
Let's not add unrelated changes - the PR is already quite big

snazy commented on versioned/spi/src/main/java/org/projectnessie/versioned/RefLogEntry.java:
```diff
@@ -0,0 +1,48 @@
+
+/** Represents a reflog-entry stored in the database. */
+@Value.Immutable
+public interface RefLogEntry {
```
Rename this one to avoid fully-qualified class names - clashes with `RefLogEntry` in `persist.proto` and  `org.projectnessie.model.RefLogResponse.RefLogEntry`

snazy reviewed (changes requested):
Still -1 on having the "details" attribute. HTML/markdown/etc UI formatting details should definitely not leak into the REST API or even the storage.

The type name `RefLogEntry` is used for three different types - that's very confusing & requires fully-qualified type names in the code, so this should be changed. 

Rest looks mostly good. Couple of very likely common code parts should be moved to the actual base classes.

snazy reviewed (commented):
Another note: I'd really like to only merge this when the TxDatabaseAdapter PR is ready. Not great to have something in m̀ain` that doesn't work everywhere.

ajantha-bhat commented on servers/quarkus-server/src/main/java/org/projectnessie/server/authz/CelAccessChecker.java:
```diff
@@ -123,6 +124,22 @@ public void canDeleteEntity(AccessContext context, NamedRef ref, ContentKey key,
   }
 
+  @Override
+  public void canViewRefLog(AccessContext context) throws AccessControlException {
```
canPerformOpOnReference works on ref, this is independent of ref. Hence added new. 
Refactoring current method again need check specific to reflog, hence added new.

ajantha-bhat commented on model/src/main/java/org/projectnessie/error/NessieRefLogNotFoundException.java:
```diff
@@ -0,0 +1,37 @@
+package org.projectnessie.error;
+
+/** This exception is thrown when a requested reflog is not present in the store. */
+public class NessieRefLogNotFoundException extends NessieNotFoundException {
```
added the testcase

ajantha-bhat commented:
@snazy 

> Still -1 on having the "details" attribute. HTML/markdown/etc UI formatting details should definitely not leak into the REST API or even the storage.

I have removed the details and added 'sourceHashes'

> The type name RefLogEntry is used for three different types - that's very confusing & requires fully-qualified type names in the code, so this should be changed.

done. 

> Rest looks mostly good. Couple of very likely common code parts should be moved to the actual base classes.

I have replied in line why I did like that. please have a look

snazy commented on servers/quarkus-server/src/test/java/org/projectnessie/server/authz/NessieAuthorizationTestProfile.java:
```diff
@@ -71,6 +71,9 @@
           .put(
               "nessie.server.authorization.rules.allow_commits_without_entity_changes",
               "op=='COMMIT_CHANGE_AGAINST_REFERENCE' && role=='test_user2' && ref.startsWith('allowedBranch')")
+          .put(
```
Needs to be documented on application.properties, too

And probably the nessie-site as well

ajantha-bhat replied:
updated the site docs, application.properties and commit kernel doc also.

snazy reviewed (commented):
Looks much better now!

Can you run sime experiments using the benchmarking tool in versioned-persist-tests - comparing the differences for in-memory, rocks and mongo. Would be good to see the perf impact of this change. If necessary, it's fine to provision an EC2 instance for these tests, if your dev machine doesn't have enough resources.

On top: this change needs a follow-up to update the nessie web site.

ajantha-bhat commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractResteasyTest.java:
```diff
@@ -465,4 +466,32 @@ public void testGetDiff() {
   }
+
+  @Test
+  public void testGetRefLog() {
```
Here this testcase mainly focus on testing rest request where as AbstractTestRest is API based. I didn't kept scenarios here as the main intention is to test rest request with query params.

ajantha-bhat replied:
we have like this for comitt log, diff api etc. So I thought it is required to have rest easy testcase also.

ajantha-bhat commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractTestRest.java:
```diff
@@ -2318,4 +2321,148 @@ protected void unwrap(Executable exec) throws Throwable {
+
+    RefLogResponse refLogResponse = api.getRefLog().get();
+    // verify reflog entries
+    assertThat(refLogResponse.getLogEntries().get(0).getOperation()).isEqualTo("DELETE_REFERENCE");
```
you mean by building an `expectedResponse` and compare `assertThat(refLogResponse).isEqualTo(expectedResponse) `? 

It has eight fields, the commit time will be set inside of the code. So, I am not sure how to set that in my expectedResponse. Hence I followed this approach.

snazy replied:
`.extracting()` or `.matches*()`

snazy replied:
But happy to get it fixed it in a follow-up

ajantha-bhat commented on versioned/persist/serialize/src/main/proto/persist.proto:
```diff
@@ -77,10 +77,34 @@ message GlobalStateLogEntry {
   repeated ContentIdWithBytes puts = 4;
 }
 
+message RefLogEntry {
```
done

ajantha-bhat commented on versioned/persist/tx/src/main/java/org/projectnessie/versioned/persist/tx/TxDatabaseAdapter.java:
```diff
@@ -243,7 +248,22 @@ public Hash merge(
-            return tryMoveNamedReference(conn, toBranch, currentHead, toHead);
+            Hash resultHash = tryMoveNamedReference(conn, toBranch, currentHead, toHead);
+
+            Hash parentId = getRefLogHead(conn);
```
done

ajantha-bhat commented on versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTransactionalDatabaseAdapter.java:
```diff
@@ -853,4 +967,71 @@ protected abstract GlobalStateLogEntry fetchFromGlobalLog(
+
+  protected Hash writeRefLogEntry(
+      NonTransactionalOperationContext ctx,
+      String refName,
```
done

snazy commented on versioned/persist/tx/src/main/java/org/projectnessie/versioned/persist/tx/TxDatabaseAdapter.java:
```diff
@@ -1304,4 +1422,167 @@ protected boolean metadataUpperCase() {
+      List<Hash> sourceHashes)
+      throws ReferenceConflictException {
+
+    Hash currentRefLogId = randomHash();
```
Nit: Looks like we don't need to generate the `currentRefLogId` but can be moved into `buildRefLogEntry`

ajantha-bhat replied:
I wanted to save one `Hash.of` conversion

ajantha-bhat replied:
moved.

snazy commented on versioned/persist/tx/src/main/java/org/projectnessie/versioned/persist/tx/TxDatabaseAdapter.java:
```diff
@@ -254,6 +269,20 @@ public Hash merge(
     }
   }
 
+  private void commitRefLog(
```
Nit: Move this method to the other new private methods for the ref-log below (try to keep pulic/protected implementation at the top - and private below)

ajantha-bhat replied:
done

snazy commented on versioned/spi/src/main/java/org/projectnessie/versioned/RefLog.java:
```diff
@@ -0,0 +1,51 @@
+  Hash getCommitHash();
+
+  /** Parent reflog id of the current entry. */
+  List<Hash> getParents();
```
Refactor to `Hash getParent()` (that list is internal to the database-adapter - nothing for the version-store)

ajantha-bhat replied:
done, Added one more interface. I was reusing it to avoid extra object.

snazy commented on versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java:
```diff
@@ -1334,4 +1336,33 @@ protected void checkExpectedGlobalStates(
+  protected Stream<RefLog> readRefLogStream(OP_CONTEXT ctx, Hash initialHash)
+      throws RefLogNotFoundException {
+    Spliterator<RefLog> split = readRefLog(ctx, initialHash);
+    return StreamSupport.stream(split, false);
```
This requires an  onClose` on the stream to delegate the close to the underlying spliterator

ajantha-bhat replied:
I have already handled for Txn store, non-txn it is not present similar to `readCommitLogStream` because there is no releaseConnection impl.

snazy replied:
Ah, right.

snazy commented on versioned/persist/tx/src/main/java/org/projectnessie/versioned/persist/tx/TxDatabaseAdapter.java:
```diff
@@ -591,6 +688,21 @@ protected int entitySize(KeyWithType entry) {
   }
 
+  @Override
+  public Stream<RefLog> refLog(Hash offset) throws RefLogNotFoundException {
```
This method should be in the "block" with the api implementations above

ajantha-bhat replied:
done

snazy reviewed (changes requested):
Pretty close.
Just need to get the list of parents out of the VS-SPI type, a missing onClose-delegation and a few nits.

ajantha-bhat commented:
@snazy : Please find the benchmarks attached. Run locally on my machine.
# JMH version: 1.34
# VM version: JDK 11.0.11, Java HotSpot(TM) 64-Bit Server VM, 11.0.11+9-LTS-194
# VM invoker: /Library/Java/JavaVirtualMachines/jdk-11.0.11.jdk/Contents/Home/bin/java
# VM options: <none>
# Blackhole mode: full + dont-inline hint (auto-detected, use -Djmh.blackhole.autoDetect=false to disable)
# Warmup: 2 iterations, 2000 ms each
# Measurement: 3 iterations, 5000 ms each
# Timeout: 10 min per iteration
# Threads: 4 threads, will synchronize iterations
# Benchmark mode: Throughput, ops/time
# Benchmark: org.projectnessie.versioned.persist.benchmarks.CommitBench.singleBranchUnsharedKeys
# Parameters: (adapter = RocksDB, tablesPerCommit = 5)

<img width="1251" alt="Screenshot 2022-01-10 at 8 46 33 PM" src="https://user-images.githubusercontent.com/5889404/148790307-ce548a13-56e4-461e-bb94-6338131e3cf0.png">
<img width="1251" alt="Screenshot 2022-01-10 at 8 46 43 PM" src="https://user-images.githubusercontent.com/5889404/148790319-59820318-ccd2-4a00-88fe-0df8390a8b4d.png">

ajantha-bhat commented:
> Pretty close.
Just need to get the list of parents out of the VS-SPI type, a missing onClose-delegation and a few nits.

Done. Please have a look. Thanks.

snazy reviewed (approved):
LGTM! Good job!

The perf regression is expected due to the nature of this change, but those don't look too bad to me.

ajantha-bhat commented:
@snazy : Thanks for the approval. I will refactor the test case in followup PR. I will merge this.
